### PR TITLE
docs: add the usage for the join order hint

### PR DESCRIPTION
**Pull request by Reminiscent**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v6.1 (TiDB 6.1 versions)
- [ ] v6.0 (TiDB 6.0 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: optimizer-hints.md [hunk 1]
```diff
@@ -244,6 +244,31 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+
+```sql
+SELECT /*+ LEADING(t1, t2) */ * FROM t1，t2, t3 WHERE t1.id = t2.id and t2.id = t3.id;
+```
+
+`LEADING()` 中表出现的顺序决定了优化器将会先对表 `t1` 和 `t2` 进行连接，在将其结果和表 `t3` 进行连接。该 Hint 比 [`STRAIGHT_JOIN`](#STRAIGHT_JOIN) 更为通用。
+
+`LEADING` 在以下情况下将会失效：
+
+* 指定了多个 leading hint；
```
Comment (qw4990) [inline optimizer-hints.md:261]:
```suggestion
1 指定了多个 leading hint；
```

File: optimizer-hints.md [hunk 2]
```diff
@@ -244,6 +244,31 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+2. leading 中指定的表名不存在；
+3. 1 个或多个 leading 中指定的表名重复；
+4. 优化器无法按照 leading 指定的顺序进行连接；
+5. 已经存在 straight_join() 的 hint；
+
```
Comment (qw4990) [inline optimizer-hints.md:266]: outer join 和 join type hint 的冲突限制也补充一下吧

File: optimizer-hints.md [hunk 3]
```diff
@@ -244,6 +244,31 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+> **注意：**
+>
+> * 无法表示 bushy tree 的连接方式。
+> * 无法固定 Join 算子的 Build/ Probe 端。
```
Comment (qw4990) [inline optimizer-hints.md:270]: 这个注意要不先不写，怕反而让用户困惑，估计大部分用户不知道 bushy tree；然后这里说无法固定 build/probe，用户配合 join type hint 一试，又有冲突，比较尴尬。。

File: optimizer-hints.md (hunk 2 above)
Comment (qw4990) [reply optimizer-hints.md:266]: 然后可以找 1-2 个例子加一下吧，提示用户如果没有生效，会有对应的 warning

**Comment by Reminiscent** [comment]
@qw4990 Update. PTAL

File: optimizer-hints.md [hunk 4]
```diff
@@ -244,6 +244,49 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+```sql
+SELECT /*+ LEADING(t1, t2) */ * FROM t1, t2, t3 WHERE t1.id = t2.id and t2.id = t3.id;
+```
+
+`LEADING()` 中表出现的顺序决定了优化器将会先对表 `t1` 和 `t2` 进行连接，在将其结果和表 `t3` 进行连接。该 Hint 比 [`STRAIGHT_JOIN`](#STRAIGHT_JOIN) 更为通用。
+
+`LEADING` 在以下情况下将会失效：
+
+1. 指定了多个 leading hint；
+
```
Comment (Reminiscent) [inline optimizer-hints.md:262]:
```suggestion
```

**Review by qw4990 — APPROVED** [review]
LGTM

File: optimizer-hints.md [hunk 5]
```diff
@@ -244,6 +244,48 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
 
 - 如果查询有除了全表扫以外的单索引扫描方式可以选择，优化器不会选择 index merge；
 
+### LEADING(t1_name [, tl_name ...])
+
+`LEADING(t1_name [, tl_name ...])` 提示优化器在生成多表连接的执行计划时会按照 Hint 中表名出现的顺序来确定多表连接的顺序。例如：
```
Comment (TomShawn) [inline optimizer-hints.md:249]:
```suggestion
`LEADING(t1_name [, tl_name ...])` 提示优化器在生成多表连接的执行计划时，按照 hint 中表名出现的顺序来确定多表连接的顺序。例如：
```

File: optimizer-hints.md [hunk 6]
```diff
@@ -244,6 +244,48 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+
+{{< copyable "sql" >}}
+
+```sql
+SELECT /*+ LEADING(t1, t2) */ * FROM t1, t2, t3 WHERE t1.id = t2.id and t2.id = t3.id;
+```
+
+`LEADING()` 中表出现的顺序决定了优化器将会先对表 `t1` 和 `t2` 进行连接，在将其结果和表 `t3` 进行连接。该 Hint 比 [`STRAIGHT_JOIN`](#STRAIGHT_JOIN) 更为通用。
```
Comment (TomShawn) [inline optimizer-hints.md:257]:
```suggestion
在以上多表连接查询语句中，`LEADING()` 中表出现的顺序决定了优化器将会先对表 `t1` 和 `t2` 进行连接，再将结果和表 `t3` 进行连接。该 hint 比 [`STRAIGHT_JOIN`](#STRAIGHT_JOIN) 更为通用。
```

File: optimizer-hints.md [hunk 7]
```diff
@@ -244,6 +244,48 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+
+`LEADING` 在以下情况下将会失效：
```
Comment (TomShawn) [inline optimizer-hints.md:259]:
```suggestion
`LEADING` hint 在以下情况下会失效：
```

File: optimizer-hints.md [hunk 8]
```diff
@@ -244,6 +244,48 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+
+1. 指定了多个 leading hint；
+2. leading 中指定的表名不存在；
+3. 1 个或多个 leading 中指定的表名重复；
+4. 优化器无法按照 leading 指定的顺序进行连接；
+5. 已经存在 straight_join() 的 hint；
+6. 查询语句中包含 outer join;
+7. 和选择 join 算法的 hint 同时使用时;
```
Comment (TomShawn) [inline optimizer-hints.md:267]: - 「1 个或多个 `LEADING` hint 中指定的表名重复」，不能指定多个 hint 吧？
- 「和选择 join 算法的 hint 同时使用时」，选择 join 算法的 hint 有哪些呢？
```suggestion
+ 指定了多个 `LEADING` hint
+ `LEADING` hint 中指定的表名不存在
+ 1 个或多个 `LEADING` hint 中指定的表名重复
+ 优化器无法按照 `LEADING` hint 指定的顺序进行表连接
+ 已经存在 `straight_join()` hint
+ 查询语句中包含 outer join
+ 和选择 join 算法的 hint 同时使用时
```

File: optimizer-hints.md [hunk 9]
```diff
@@ -244,6 +244,48 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+
+当出现了上述有冲突的情况，会输出 warning 警告。
```
Comment (TomShawn) [inline optimizer-hints.md:269]:
```suggestion
当出现了上述失效的情况，会输出 warning 警告。
```

File: optimizer-hints.md [hunk 10]
```diff
@@ -244,6 +244,48 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+
+```sql
+-- 指定了多个 leading hint
```
Comment (TomShawn) [inline optimizer-hints.md:272]:
```suggestion
-- 指定了多个 LEADING hint
```

File: optimizer-hints.md [hunk 11]
```diff
@@ -244,6 +244,48 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+
+SELECT /*+ LEADING(t1, t2) LEADING(t3) */ * FROM t1, t2, t3 WHERE t1.id = t2.id and t2.id = t3.id;
+
+-- 通过执行 `show warnings` 了解具体产生冲突的原因
+
+show warnings;
```
Comment (TomShawn) [inline optimizer-hints.md:278]:
```suggestion
SHOW WARNINGS;
```

File: optimizer-hints.md [hunk 12]
```diff
@@ -244,6 +244,48 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+```
+```sql
++---------+------+-------------------------------------------------------------------------------------------------------------------+
+| Level   | Code | Message                                                                                                           |
++---------+------+-------------------------------------------------------------------------------------------------------------------+
+| Warning | 1815 | We can only use one leading hint at most, when multiple leading hints are used, all leading hints will be invalid |
++---------+------+-------------------------------------------------------------------------------------------------------------------+
+
```
Comment (TomShawn) [inline optimizer-hints.md:286]:
```suggestion
```

**Review by TomShawn — COMMENTED** [review]
mostly minor comments

File: optimizer-hints.md [hunk 13]
```diff
@@ -244,6 +244,47 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+SELECT /*+ LEADING(t1, t2) */ * FROM t1, t2, t3 WHERE t1.id = t2.id and t2.id = t3.id;
+```
+
+在以上多表连接查询语句中，`LEADING()` 中表出现的顺序决定了优化器将会先对表 `t1` 和 `t2` 进行连接，再将结果和表 `t3` 进行连接。该 hint 比 [`STRAIGHT_JOIN`](#STRAIGHT_JOIN) 更为通用。
+
+`LEADING` hint 在以下情况下会失效：
+
++ 指定了多个 `LEADING` hint
++ `LEADING` hint 中指定的表名不存在
++ 1 个或多个 `LEADING` hint 中指定的表名重复
```
Comment (Reminiscent) [inline optimizer-hints.md:263]:
```suggestion
+ `LEADING` hint 中指定了重复的表名
```

File: optimizer-hints.md (hunk 8 above)
Comment (Reminiscent) [reply optimizer-hints.md]: 第二个问题包括：MERGE_JOIN INL_JOIN INL_HASH_JOIN HASH_JOIN 这四个，要一起加上去么。
第一个问题改动方法提到下面了

File: optimizer-hints.md (hunk 13 above)
Comment (Reminiscent) [reply optimizer-hints.md:263]: @TomShawn 麻烦帮忙看看描述

File: optimizer-hints.md [hunk 14]
```diff
@@ -244,6 +244,47 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+
+`LEADING` hint 在以下情况下会失效：
+
++ 指定了多个 `LEADING` hint
++ `LEADING` hint 中指定的表名不存在
++ `LEADING` hint 中指定了重复的表名
++ 优化器无法按照 `LEADING` hint 指定的顺序进行表连接
++ 已经存在 `straight_join()` hint
++ 查询语句中包含 outer join
++ 和选择 join 算法的 hint 同时使用时
```
Comment (TomShawn) [inline optimizer-hints.md:267]:
```suggestion
+ 和选择 join 算法的 hint（即 `MERGE_JOIN`、`INL_JOIN`、`INL_HASH_JOIN`、`HASH_JOIN`）同时使用时
```

**Review by TomShawn — APPROVED** [review]
LGTM

File: optimizer-hints.md (hunk 8 above)
Comment (TomShawn) [reply optimizer-hints.md]: 已添加

File: optimizer-hints.md [hunk 15]
```diff
@@ -244,6 +244,47 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+
+`LEADING(t1_name [, tl_name ...])` 提示优化器在生成多表连接的执行计划时，按照 hint 中表名出现的顺序来确定多表连接的顺序。例如：
+
+{{< copyable "sql" >}}
+
+```sql
+SELECT /*+ LEADING(t1, t2) */ * FROM t1, t2, t3 WHERE t1.id = t2.id and t2.id = t3.id;
+```
+
+在以上多表连接查询语句中，`LEADING()` 中表出现的顺序决定了优化器将会先对表 `t1` 和 `t2` 进行连接，再将结果和表 `t3` 进行连接。该 hint 比 [`STRAIGHT_JOIN`](#STRAIGHT_JOIN) 更为通用。
```
Comment (TomShawn) [inline optimizer-hints.md:257]:
```suggestion
在以上多表连接查询语句中，`LEADING()` 中表出现的顺序决定了优化器将会先对表 `t1` 和 `t2` 进行连接，再将结果和表 `t3` 进行连接。该 hint 比 [`STRAIGHT_JOIN`](#straight_join) 更为通用。
```

File: optimizer-hints.md [hunk 16]
```diff
@@ -244,6 +244,47 @@ SELECT /*+ USE_INDEX_MERGE(t1, idx_a, idx_b, idx_c) */ * FROM t1 WHERE t1.a > 10
...
+```sql
+-- 指定了多个 LEADING hint
+
+SELECT /*+ LEADING(t1, t2) LEADING(t3) */ * FROM t1, t2, t3 WHERE t1.id = t2.id and t2.id = t3.id;
+
+-- 通过执行 `show warnings` 了解具体产生冲突的原因
+
+SHOW WARNINGS;
+```
+```sql
```
Comment (TomShawn) [inline optimizer-hints.md:280]:
````suggestion

```sql
````

**Pull request merged**